From b7ac6f175f92b5bdf758fb31671922348afd51af Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:16:27 -0400
Subject: [net] netfilter: nf_tables: add nft_dereference() macro

Message-id: <20141021161627.22888.73818.stgit@localhost.localdomain>
Patchwork-id: 97915
O-Subject: [RHEL7 net PATCH 018/124] netfilter: nf_tables: add nft_dereference() macro
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 67a8fc27cca06e185c1ab39baaccd2103f6f9f51

 commit 67a8fc27cca06e185c1ab39baaccd2103f6f9f51
 Author: Patrick McHardy <kaber@trash.net>
 Date:   Tue Feb 18 18:06:49 2014 +0000

    netfilter: nf_tables: add nft_dereference() macro

    Signed-off-by: Patrick McHardy <kaber@trash.net>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/net/netfilter/nf_tables.h b/include/net/netfilter/nf_tables.h
index e7e14ff..81abd61 100644
--- a/include/net/netfilter/nf_tables.h
+++ b/include/net/netfilter/nf_tables.h
@@ -3,6 +3,7 @@
 
 #include <linux/list.h>
 #include <linux/netfilter.h>
+#include <linux/netfilter/nfnetlink.h>
 #include <linux/netfilter/x_tables.h>
 #include <linux/netfilter/nf_tables.h>
 #include <net/netlink.h>
@@ -521,6 +522,9 @@ void nft_unregister_chain_type(const struct nf_chain_type *);
 int nft_register_expr(struct nft_expr_type *);
 void nft_unregister_expr(struct nft_expr_type *);
 
+#define nft_dereference(p)     \
+ nfnl_dereference(p, NFNL_SUBSYS_NFTABLES)
+
 #define MODULE_ALIAS_NFT_FAMILY(family) \
  MODULE_ALIAS("nft-afinfo-" __stringify(family))
 
diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index adce01e..4b7e14d 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -794,9 +794,8 @@ nf_tables_counters(struct nft_base_chain *chain, const struct nlattr *attr)
  stats->pkts = be64_to_cpu(nla_get_be64(tb[NFTA_COUNTER_PACKETS]));
 
  if (chain->stats) {
-  /* nfnl_lock is held, add some nfnl function for this, later */
   struct nft_stats __percpu *oldstats =
-   rcu_dereference_protected(chain->stats, 1);
+    nft_dereference(chain->stats);
 
   rcu_assign_pointer(chain->stats, newstats);
   synchronize_rcu();
-- 
1.7.1